tests/test_bounce_integral.py

```diff
@@ -0,0 +1,751 @@
+    y = _affine_bijection_forward(x, a, b)
+    x_1 = affine_bijection(y, a, b)
+    np.testing.assert_allclose(x_1, x)
+    np.testing.assert_allclose(_affine_bijection_forward(x_1, a, b), y)
+    np.testing.assert_allclose(automorphism_arcsin(automorphism_sin(y)), y, atol=5e-7)
+    np.testing.assert_allclose(automorphism_sin(automorphism_arcsin(y)), y, atol=5e-7)
+
+    np.testing.assert_allclose(grad_affine_bijection(a, b), 1 / (2 / (b - a)))
+    np.testing.assert_allclose(
+        grad_automorphism_sin(y),
+        1 / grad_automorphism_arcsin(automorphism_sin(y)),
+        atol=2e-6,
+    )
+    np.testing.assert_allclose(
+        1 / grad_automorphism_arcsin(y),
+        grad_automorphism_sin(automorphism_arcsin(y)),
+        atol=2e-6,
+    )
+
+    # test that floating point error is acceptable
+    x = tanh_sinh(19)[0]
+    assert np.all(np.abs(x) < 1)
+    y = 1 / np.sqrt(1 - np.abs(x))
+    assert np.isfinite(y).all()
+    y = 1 / np.sqrt(1 - np.abs(automorphism_sin(x)))
+    assert np.isfinite(y).all()
+    y = 1 / np.sqrt(1 - np.abs(automorphism_arcsin(x)))
+    assert np.isfinite(y).all()
+
+
+@pytest.mark.unit
+def test_bounce_quadrature():
+    """Test bounce integral matches elliptic integral."""
+    p = 1e-4
+    m = 1 - p
+    # Some prime number that doesn't appear anywhere in calculation.
+    # Ensures no lucky cancellation occurs from this test case since otherwise
+    # (bp2 - bp1) / pi = pi / (bp2 - bp1) which could mask errors since pi
+    # appears often in transformations.
+    v = 7
+    truth = v * 2 * ellipkm1(p)
+    rtol = 1e-4
+
+    def integrand(B, pitch):
+        return 1 / jnp.sqrt(1 - pitch * m * B)
+
+    bp1 = -np.pi / 2 * v
+    bp2 = -bp1
+    knots = np.linspace(bp1, bp2, 50)
+    B = np.clip(np.sin(knots / v) ** 2, 1e-7, 1)
+    B_z_ra = np.sin(2 * knots / v) / v
+    pitch = 1 + 50 * jnp.finfo(jnp.array(1.0).dtype).eps
+
+    bounce_integrate, _ = bounce_integral(
+        B, B, B_z_ra, knots, quad=tanh_sinh(40), automorphism=None, check=True
+    )
+    tanh_sinh_vanilla = _filter_not_nan(bounce_integrate(integrand, [], pitch))
+    assert tanh_sinh_vanilla.size == 1
+    np.testing.assert_allclose(tanh_sinh_vanilla, truth, rtol=rtol)
+
+    bounce_integrate, _ = bounce_integral(
+        B, B, B_z_ra, knots, quad=leggauss(25), check=True
+    )
+    leg_gauss_sin = _filter_not_nan(bounce_integrate(integrand, [], pitch))
+    assert leg_gauss_sin.size == 1
+    np.testing.assert_allclose(leg_gauss_sin, truth, rtol=rtol)
+
+
+@pytest.mark.unit
+def test_bounce_integral_checks():
+    """Test that all the internal correctness checks pass for real example."""
+    # Suppose we want to compute a bounce average of the function
+    # f(ℓ) = (1 − λ |B|) * g_zz, where g_zz is the squared norm of the
+    # toroidal basis vector on some set of field lines specified by (ρ, α)
+    # coordinates. This is defined as
+    # (∫ f(ℓ) / √(1 − λ |B|) dℓ) / (∫ 1 / √(1 − λ |B|) dℓ)
+    eq = get("HELIOTRON")
```

Comment:
It's nice that you have added an actual equilibrium. Is there a way to compare it with the actual truth?

Reply:
Not that I know of. There are other numerical methods we can test against if we want. With sufficient knot density the spline of $|B|$ reconstructs $|B|$ to machine precision, so we could give the spline to scipy ODE integrate and solve dK/dℓ = f(ℓ) / √(1 − λ |B|) on [bounce point left, bounce point right] K(bounce_point left) = 0. Likewise with the denominator.